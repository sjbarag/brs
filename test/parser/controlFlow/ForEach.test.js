const BrsError = require("../../../lib/Error");
const { Lexeme, Parser } = require("brs");
const { Expr, Stmt } = Parser;

const { EOF } = require("../ParserTests");

describe("parser foreach loops", () => {
    afterEach(() => BrsError.reset());

    it("requires a name and target", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.ForEach, text: "for each", line: 2 },
            { kind: Lexeme.Identifier, text: "word", line: 2 },
            { kind: Lexeme.Identifier, text: "in", line: 2 },
            { kind: Lexeme.Identifier, text: "lipsum", line: 2 },
            { kind: Lexeme.Newline, text: "\n", line: 2 },

            // body would go here, but it's not necessary for this test
            { kind: Lexeme.EndFor, text: "end for", line: 3 },
            { kind: Lexeme.Newline, text: "\n", line: 3 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();

        let forEach = parsed[0];
        expect(forEach).toBeInstanceOf(Stmt.ForEach);

        expect(forEach.item).toEqual(
            { kind: Lexeme.Identifier, text: "word", line: 2 }
        );
        expect(forEach.target).toBeInstanceOf(Expr.Variable);
        expect(forEach.target.name).toEqual(
            { kind: Lexeme.Identifier, text: "lipsum", line: 2 }
        );

        expect(parsed).toMatchSnapshot();
    });
});
