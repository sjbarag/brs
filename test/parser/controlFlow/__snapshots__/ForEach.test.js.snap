// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser foreach loops allows 'next' to terminate loop 1`] = `
Array [
  ForEach {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [],
      "type": "Block",
    },
    "item": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "word",
    },
    "target": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "lipsum",
      },
      "type": "Variable",
    },
    "tokens": Object {
      "endFor": Object {
        "isReserved": true,
        "kind": "Next",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "next",
      },
      "forEach": Object {
        "isReserved": false,
        "kind": "ForEach",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "for each",
      },
      "in": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "in",
      },
    },
    "type": "ForEach",
  },
]
`;

exports[`parser foreach loops requires a name and target 1`] = `
Array [
  ForEach {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [],
      "type": "Block",
    },
    "item": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "word",
    },
    "target": Variable {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "lipsum",
      },
      "type": "Variable",
    },
    "tokens": Object {
      "endFor": Object {
        "isReserved": false,
        "kind": "EndFor",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end for",
      },
      "forEach": Object {
        "isReserved": false,
        "kind": "ForEach",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "for each",
      },
      "in": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "in",
      },
    },
    "type": "ForEach",
  },
]
`;
